### PR TITLE
DFUSE-2276 dfuse: Launch and start a FUSE filesystem on startup.

### PR DESCRIPTION
**Pull request by ashleypittman**
Write a getopt() routine to accept pool/container options.
Simplify the structs used by dfuse_core.c and dfuse_main.c to
remove duplicate entries now only one FUSE fs is required.
Change dfuse_fs_send() to return an error.
Change types of argunments to functions where sensible.

This allows a DFUSE to connect to a DAOS container, and create
a FUSE filesystem.  Any activity on the filesystem currently
returns EIO.

### CODE REVIEW

File: src/client/dfuse/ops/mkdir.c [hunk 1]
```diff
@@ -24,41 +24,65 @@
...
+	D_ALLOC_PTR(inode);
+	if (!inode) {
+		D_GOTO(err, rc = ENOMEM);
+	}
+
+	DFUSE_TRA_INFO(parent_inode, "parent");
+
+	/* mkdir with the correct parent */
+	rc = dfs_mkdir(fs_handle->fsh_dfs, parent_inode->obj, name, mode);
+	if (rc != -DER_SUCCESS) {
```
Comment (mchaarawi) [inline src/client/dfuse/ops/mkdir.c]: i have in dfuse_hl something like a daos error code to errno code convertor. not sure if you are doing something similar, but if not you can reuse that.

File: src/client/dfuse/ops/mkdir.c [hunk 2]
```diff
@@ -24,41 +24,65 @@
...
+	}
 
-	desc->request.ir_inode_num = parent;
+	strncpy(inode->name, name, NAME_MAX);
+	inode->parent = parent;
+	atomic_fetch_add(&inode->ie_ref, 1);
 
-	rc = dfuse_fs_send(&desc->request);
-	if (rc != 0)
+	rc = dfs_lookup_rel(fs_handle->fsh_dfs, parent_inode->obj, name,
```
Comment (mchaarawi) [inline src/client/dfuse/ops/mkdir.c]: why do we lookup after we create the directory? if mkdir here needs to also return an open handle on the directory, maybe then use dfs_open with dir type?

File: src/client/dfuse/ops/mkdir.c [hunk 3]
```diff
@@ -24,41 +24,65 @@
...
+			    O_RDONLY, &inode->obj, &mode);
+	if (rc != -DER_SUCCESS) {
 		D_GOTO(err, 0);
+	}
+
+	rc = dfs_ostat(fs_handle->fsh_dfs, inode->obj, &inode->stat);
```
Comment (mchaarawi) [inline src/client/dfuse/ops/mkdir.c]: also not sure why we need to ostat

File: src/client/dfuse/ops/lookup.c [hunk 4]
```diff
@@ -21,95 +21,102 @@
...
+			D_GOTO(out_decref, rc = ENOENT);
+		} else {
+			D_GOTO(out_decref, rc = EIO);
+		}
+	}
+
+	strncpy(inode->name, name, NAME_MAX);
+	atomic_fetch_add(&inode->ie_ref, 1);
+
+	rc = dfs_ostat(fs_handle->fsh_dfs, inode->obj, &inode->stat);
```
Comment (mchaarawi) [inline src/client/dfuse/ops/lookup.c]: why do we ostat after a lookup?

File: src/client/dfs/dfs.c [hunk 5]
```diff
@@ -2862,6 +2863,12 @@ dfs_listxattr(dfs_t *dfs, dfs_obj_t *obj, char *list, daos_size_t *size)
 	return rc;
 }
 
+daos_obj_id_t
+dfs_obj2id(dfs_obj_t *obj)
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfs/dfs.c]: I didn't realise that daos_obj_id_t was a struct when I wrote this, but yes you can copy structs with assignment.  I've changed the code to be consistent with other functions however.

File: src/client/dfuse/dfuse.h [hunk 6]
```diff
@@ -30,85 +30,59 @@
...
+	char	*svcl;
+	char	*group;
+	char	*mountpoint;
+	bool	threaded;
 };
 
 struct dfuse_info {
-	struct dfuse_state	*dfuse_state;
-	struct fs_info		ci_fsinfo;
+	struct fuse_session		*fsi_session;
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse.h]: This struct was previously called fs_info, I've changed the prefix to dfi_ so it's closer to the struct name however.

File: src/client/dfuse/dfuse.h [hunk 7]
```diff
@@ -30,85 +30,59 @@
...
-	char				*dfuse_prefix;
-	/** ctrl_fs inoss directory handle */
-	struct ctrl_dir			*ionss_dir;
-	/** ctrl_fs projections directory handle */
-	struct ctrl_dir			*projections_dir;
-	/** Group information */
-	struct dfuse_service_group	grp;
-};
-
 struct dfuse_projection_info {
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse.h]: You're commenting on removed lines, after this commit all entries are consistent, although as above I've changed the name now as well.

File: src/client/dfuse/dfuse.h [hunk 8]
```diff
@@ -30,85 +30,59 @@
...
-void
-dfuse_reg(struct dfuse_state *dfuse_state, struct dfuse_info *dfuse_info);
-
-void
-dfuse_post_start(struct dfuse_state *dfuse_state);
+		   struct dfuse_projection_info *fsi_handle);
 
-void
-dfuse_finish(struct dfuse_state *dfuse_state);
+int
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse.h]: Done.  I've renamed a couple of these functions, removed one entirely and added comments to the rest.

File: src/client/dfuse/dfuse_core.c [hunk 9]
```diff
@@ -342,16 +342,9 @@ wb_reset(void *arg)
...
-void
-dfuse_reg(struct dfuse_state *dfuse_state, struct dfuse_info *dfuse_info)
-{
-	dfuse_state->dfuse_info = dfuse_info;
-}
-
-static bool
-initialize_projection(struct dfuse_state *dfuse_state)
+int
+dfuse_post_start(struct dfuse_info *dfuse_info)
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/dfuse_core.c]: There are some unused types here, but we don't know which ones are going to be used when the callbacks are ported across, so I prefer to remove the initialisation code, and the type itself at the same time of that's the correct thing for the specific callback.

File: src/client/dfuse/ops/lookup.c [hunk 10]
```diff
@@ -21,95 +21,102 @@
...
-	struct dfuse_entry_out		*out = request->out;
 	struct fuse_entry_param		entry = {0};
 	d_list_t			*rlink;
-	bool				keep_ref = false;
+	daos_obj_id_t oid = dfs_obj2id(inode->obj);
 
-	DFUSE_REQUEST_RESOLVE(request, out);
-	if (request->rc)
-		D_GOTO(out, 0);
+	inode->stat.st_ino = (ino_t)oid.hi;
```
[the remark this answers is not in the document]
Comment (ashleypittman) [reply src/client/dfuse/ops/lookup.c]: I suppose just using lo would have the same problem?  Lustre manages this by ORing the two halves together, although there's scope for conflict here still.

The plan is to add a indirection table here to lookup DAOS IDs to inode numbers which is required for multiple containers in the same filesystem, I'm hoping to work on that next week so we might have to live with this for now until it's addressed properly.

File: src/client/dfuse/ops/mkdir.c (hunk 1 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/mkdir.c]: Yes, this is a problem and was never easy to get right in IOF.  Currently DFUSE_REPLY_ERR_RAW converts negative numbers to EIO, but this is a blunt hammer.

The error_convert() function in dfuse_hl.c seems to return a mixture of CaRT and system error numbers, and to compound the issue always negative values.

I think FUSE high level uses negative error numbers as there cases where positive numbers are used for size and negative numbers for errors, however FUSE low level avoids this and correctly uses positive error numbers in all cases.

File: src/client/dfuse/ops/mkdir.c (hunk 2 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/mkdir.c]: Do you mean with a combination of O_CREAT and S_IFDIR?, yes that might be better.  I'll take a look later on today, but push a PR with the simpler issues resolved first.

File: src/client/dfuse/ops/mkdir.c (hunk 3 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/mkdir.c]: After a successful mkdir we need to call fuse_reply_entry(), which passes a struct stat from the newly created directory back to the kernel.  If you lookup what fuse high level is doing it'll always call getattr() after mkdir() to know what values to reply.

File: src/client/dfuse/ops/lookup.c (hunk 4 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/lookup.c]: Same as mkdir, the reply to lookup is fuse_reply_entry() which passes a stat call into the kernel.

File: src/client/dfuse/ops/mkdir.c (hunk 2 above)
Comment (ashleypittman) [reply src/client/dfuse/ops/mkdir.c]: This was quite easy, tested and pushed in a later commit.

File: src/client/dfuse/ops/lookup.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfuse/ops/lookup.c]: OK sounds good. just indicating that this won't work.

**Comment by johannlombardi** [comment]
Many failures. Do we know what they are about? Cannot land as is, i am afraid.

**Comment by ashleypittman** [comment]
> Many failures. Do we know what they are about?

The build failure is general CI issues, I don't think anything is building at the minute.

> Cannot land as is, i am afraid.

Because of the CI failures or anything else?  Do I need to work on this or should I wait until CI is fixed?

File: src/client/dfuse/dfuse_core.c [hunk 11]
```diff
@@ -76,6 +57,7 @@ dfuse_fs_send(struct dfuse_request *request)
 			request->ir_ht = RHS_INODE;
 		}
 	}
+	return EIO;
```
Comment (johannlombardi) [inline src/client/dfuse/dfuse_core.c]: I see two returns there. Merge issue?

**Pull request merged**